File: backend/src/main/java/dev/adriankluge/hobbyhub/mtg/meta/client/EdhrecClient.java
package dev.adriankluge.hobbyhub.mtg.meta.client;

import java.util.List;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

// EDHREC's open, key-free JSON API (json.edhrec.com) - the same data source
// that powers edhrec.com's own pages. Used for two of the four "Meta &
// Stats" widgets (section 4.5): most-played cards and popular Commander
// decks. Timeout comes from the shared RestClientCustomizer (see
// config.RestClientConfig), same as every other RestClient this app builds.
@Component
public class EdhrecClient {

    private static final int MAX_ITEMS = 3;

    private final RestClient restClient;

    public EdhrecClient(RestClient.Builder restClientBuilder, @Value("${app.mtg.edhrec-json-base-url}") String baseUrl) {
        this.restClient = restClientBuilder.baseUrl(baseUrl).defaultHeader("Accept", "application/json").build();
    }

    public List<MetaEntry> mostPlayedCards() {
        return fetch("/pages/top/week.json");
    }

    public List<MetaEntry> popularCommanderDecks() {
        return fetch("/pages/commanders/week.json");
    }

    private List<MetaEntry> fetch(String path) {
        EdhrecPageDto response = restClient.get().uri(path).retrieve().body(EdhrecPageDto.class);
        if (response == null) {
            return List.of();
        }
        return response.firstCardlistViews().stream()
                .limit(MAX_ITEMS)
                .map(view -> new MetaEntry(view.name(), "https://edhrec.com" + view.url(), view.numDecks()))
                .toList();
    }
}

File: api/lib/EdhrecClient.php
<?php
require_once __DIR__ . '/http_client.php';

// EDHREC's open, key-free JSON API (json.edhrec.com) - the same data source
// that powers edhrec.com's own pages. Used for two of the four MTG Meta &
// Stats widgets: most-played cards and popular Commander decks.
class EdhrecClient
{
    private const MAX_ITEMS = 3;

    public function mostPlayedCards(): array
    {
        return $this->fetch('/pages/top/week.json');
    }

    public function popularCommanderDecks(): array
    {
        return $this->fetch('/pages/commanders/week.json');
    }

    private function fetch(string $path): array
    {
        $json = http_get_json(EDHREC_JSON_BASE_URL . $path);
        if ($json === null) {
            // A genuine request failure - propagate so the caller can keep
            // the existing cache (see cron/refresh_mtg_meta.php).
            throw new RuntimeException("EDHREC request failed: $path");
        }

        $cardviews = $json['container']['json_dict']['cardlists'][0]['cardviews'] ?? [];
        $items = [];
        foreach (array_slice($cardviews, 0, self::MAX_ITEMS) as $view) {
            $items[] = [
                'name' => $view['name'] ?? '',
                'url' => 'https://edhrec.com' . ($view['url'] ?? ''),
                'numDecks' => $view['num_decks'] ?? null,
            ];
        }
        return $items;
    }
}
